### PR TITLE
SQL, part II

### PR DESCRIPTION
**Pull request by MikhailArkhipov**
#1078 Drop file from Solution Explorer should yield source(file) 
``` *.r => source(~/...)```
```*.sql => 'SELECT * ...'``` (SQL file content)
other types -> string with ```~/path```

#1593 Support Add Database Connection 
 - Add Database Connection...
 - Add Data Source...
 - Manage Data Sources...

Allowed without project (generates ```dbConnection``` variable.
In project generates variable ```dbConnectionNNN``` and saves in project settings
Supports UI editor for connection strings in settings

#2064 Support creation of ODBC DSNs 
 - Basically invokes OS ODBC functionality

### CODE REVIEW

File: src/Package/Impl/Commands/SessionCommand.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using Microsoft.R.Components.InteractiveWorkflow;
+using Microsoft.R.Host.Client;
+using Microsoft.VisualStudio.R.Packages.R;
+
+namespace Microsoft.VisualStudio.R.Package.Commands {
+    internal abstract class SessionCommand : PackageCommand {
+        private readonly IRInteractiveWorkflowProvider _workflowProvider;
```
Comment (AlexanderSher) [inline src/Package/Impl/Commands/SessionCommand.cs]: This field isn't used

File: src/Package/Impl/ProjectSystem/Configuration/ProjectConfigurationSettingsProvider.cs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+        public Task<IProjectConfigurationSettingsAccess> OpenProjectSettingsAccessAsync(ConfiguredProject configuredProject) {
+            var props = configuredProject.Services.ExportProvider.GetExportedValue<ProjectProperties>();
+            return OpenProjectSettingsAccessAsync(configuredProject.UnconfiguredProject, props);
+        }
+
+        public async Task<IProjectConfigurationSettingsAccess> OpenProjectSettingsAccessAsync(
+            UnconfiguredProject unconfiguredProject, IRProjectProperties propertes) {
+
+            Check.ArgumentNull(nameof(unconfiguredProject), unconfiguredProject);
```
Comment (AlexanderSher) [inline src/Package/Impl/ProjectSystem/Configuration/ProjectConfigurationSettingsProvider.cs]: What is this `Check.ArgumentNull`? Is it implemented in our code?

File: src/Package/Impl/ProjectSystem/Configuration/ProjectConfigurationSettingsProvider.cs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+            }
+
+            private void OnCollectionChanged(object sender, NotifyCollectionChangedEventArgs e) {
+                _changed = true;
+            }
+
+            #region IProjectConfigurationSettingsAccess
+            public ConfigurationSettingCollection Settings { get; }
+            public void Dispose() {
+                if (UseCount > 0) {
```
Comment (AlexanderSher) [inline src/Package/Impl/ProjectSystem/Configuration/ProjectConfigurationSettingsProvider.cs]: You can use `CountdownDisposable`

File: src/Package/Impl/ProjectSystem/PropertyPages/Settings/SettingsPageControl.cs [hunk 4]
```diff
@@ -167,16 +188,13 @@ internal partial class SettingsPageControl : UserControl {
...
-            // Set the correct font
-            var fontSvc = VsAppShell.Current.GetGlobalService<IUIHostLocale2>(typeof(SUIHostLocale));
-            if (fontSvc != null) {
-                var logFont = new UIDLGLOGFONT[1];
-                int hr = fontSvc.GetDialogFont(logFont);
-                if (hr == VSConstants.S_OK) {
-                    this.Font = IdeUtilities.FontFromUiDialogFont(logFont[0]);
-                }
-            }
+            this.Font = (_coreShell as IApplicationShell)?.GetUiFont() ?? this.Font;
```
Comment (AlexanderSher) [inline src/Package/Impl/ProjectSystem/PropertyPages/Settings/SettingsPageControl.cs]: You can change the type of `_coreShell` to `IApplicationShell`

File: src/Package/Impl/ProjectSystem/RDebugLaunchProvider.cs [hunk 5]
```diff
@@ -93,12 +93,13 @@ public RDebugLaunchProvider(ConfiguredProject configuredProject, IRInteractiveWo
...
 
             var settingsFile = await _properties.GetSettingsFileAsync();
-
             if (!string.IsNullOrWhiteSpace(settingsFile)) {
                 var activeProject = _pss.GetActiveProject();
                 if (activeProject != null) {
                     settingsFile = settingsFile.MakeAbsolutePathFromRRelative(Path.GetDirectoryName(activeProject.FullName));
-                    await SourceFileAsync(settingsFile, Invariant($"{Resources.Launch_SettingsFileDoesNotExist} {settingsFile}"));
+                    if (FileSystem.FileExists(settingsFile)) {
+                        await SourceFileAsync(settingsFile, Invariant($"{Resources.Launch_SettingsFileDoesNotExist} {settingsFile}"));
```
Comment (AlexanderSher) [inline src/Package/Impl/ProjectSystem/RDebugLaunchProvider.cs]: We need `try/catch` for `IOException` here if `SourceFileAsync` doesn't have it

File: src/Package/Impl/Sql/Commands/ConfigurationSettingCommand.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+                using (var access = await _projectConfigurationSettingsProvider.OpenProjectSettingsAccessAsync(configuredProject)) {
+                    name = access.Settings.FindNextAvailableSettingName(_settingNameTemplate);
+                    var s = new ConfigurationSetting(name, value ?? name, ConfigurationSettingValueType.String);
+                    s.EditorType = ConnectionStringEditor.ConnectionStringEditorName;
+                    s.Category = ConnectionStringEditor.ConnectionStringEditorCategory;
+                    s.Description = Resources.ConnectionStringDescription;
+                    access.Settings.Add(s);
+                }
+            }
+            await RSession?.EvaluateAsync(Invariant($"{name ?? _settingNameTemplate} <- '{value}'"), REvaluationKind.Mutating);
```
Comment (AlexanderSher) [inline src/Package/Impl/Sql/Commands/ConfigurationSettingCommand.cs]: You should catch `MessageTransportExpection` here.
Also, do we want to show anything to user if command fails?

File: src/Package/Test/Project/ProjectSystem/PropertyPages/ProjectSettingsViewModelTest.cs [hunk 7]
```diff
@@ -70,5 +71,23 @@ public class ProjectSettingsViewModelTest {
             model.Files.ToArray()[0].Should().Be("~/Settings.R");
             model.Files.ToArray()[1].Should().Be("~/Debug.Settings.R");
         }
+
+        [Test]
+        public async Task LoadSave() {
```
Comment (AlexanderSher) [inline src/Package/Test/Project/ProjectSystem/PropertyPages/ProjectSettingsViewModelTest.cs]: Can we make a test on a real files?

File: src/Package/Impl/ProjectSystem/Configuration/ProjectConfigurationSettingsProvider.cs (hunk 2 above)
Comment (MikhailArkhipov) [inline src/Package/Impl/ProjectSystem/Configuration/ProjectConfigurationSettingsProvider.cs]: Yes, we have it in Core.Diagnostics. It is actually an old class from Blend, haven't seen much use in RTVS though

File: src/Package/Impl/Sql/Commands/ConfigurationSettingCommand.cs (hunk 6 above)
Comment (MikhailArkhipov) [inline src/Package/Impl/Sql/Commands/ConfigurationSettingCommand.cs]: Actually yes, I think it would be nice to have it visible and added to history

File: src/Package/Test/Project/ProjectSystem/PropertyPages/ProjectSettingsViewModelTest.cs (hunk 7 above)
Comment (MikhailArkhipov) [inline src/Package/Test/Project/ProjectSystem/PropertyPages/ProjectSettingsViewModelTest.cs]: ProjectSettingsControlTest uses real files

File: src/Package/Impl/ProjectSystem/Configuration/ProjectConfigurationSettingsProvider.cs (hunk 3 above)
Comment (MikhailArkhipov) [inline src/Package/Impl/ProjectSystem/Configuration/ProjectConfigurationSettingsProvider.cs]: It would have to be ```CountdownDisposable<T>``` since I need disposable object with more methods than just plain IDisposable. Or unseal it and derive SettingsAccess from it

File: src/Package/Impl/ProjectSystem/RDebugLaunchProvider.cs (hunk 5 above)
Comment (MikhailArkhipov) [inline src/Package/Impl/ProjectSystem/RDebugLaunchProvider.cs]: SourceFileAsync silences everything AFAICT

**Comment by MikhailArkhipov** [comment]
- Made setting eval visible
- assorted CR feedback
- used CountdownDisposable

File: src/Package/Impl/ProjectSystem/Configuration/ProjectConfigurationSettingsProvider.cs [hunk 8]
```diff
@@ -91,34 +89,38 @@ internal sealed class ProjectConfigurationSettingsProvider : IProjectConfigurati
...
             private readonly ProjectConfigurationSettingsProvider _provider;
+            private readonly CountdownDisposable _counter;
             private readonly string _projectPath;
             private bool _changed;
 
             public SettingsAccess(ProjectConfigurationSettingsProvider provider, string projectPath, ConfigurationSettingCollection settings) {
                 _provider = provider;
                 _projectPath = projectPath;
+
+                _counter = new CountdownDisposable(() => Release());
```
Comment (AlexanderSher) [inline src/Package/Impl/ProjectSystem/Configuration/ProjectConfigurationSettingsProvider.cs]: just `_counter = new CountdownDisposable(Release)`

**Comment by AlexanderSher** [comment]
:+1:
